src/types/User.ts

```diff
@@ -49,3 +67,83 @@ export enum UsernameVisibility {
   hover = "hover",
   inline = "inline",
 }
+
+export const ExperienceSchema: Yup.ObjectSchema<Experience> = Yup.object()
+  .shape({
+    // @debt refactor bartender to be potentially undefined. Or can we remove it entirely?
+    bartender: Yup.lazy(() => UserSchema.required()),
+    table: Yup.string().nullable(),
+    row: Yup.number().nullable(),
+    column: Yup.number().nullable(),
+  })
+  .required();
+
+export const UserExperienceDataSchema = Yup.lazy<
+  UserExperienceData | undefined
+>((data) => {
+  const lazyObjectShape = Object.fromEntries(
+    Object.keys(data ?? {}).map((key) => [key, ExperienceSchema])
+  );
+
+  return Yup.object().shape(lazyObjectShape).noUnknown();
+});
+
+export const VideoStateSchema: Yup.ObjectSchema<VideoState> = Yup.object()
+  .shape({
+    inRoomOwnedBy: Yup.string(),
+    removedParticipantUids: Yup.array().of(Yup.string().required()),
+  })
+  .required();
+
+export const MyPersonalizedScheduleSchema = Yup.lazy<
+  MyPersonalizedSchedule | undefined
+>((data) => {
+  const lazyObjectShape = Object.fromEntries(
+    Object.keys(data ?? {}).map((key) => [
+      key,
+      Yup.array().of(Yup.string().required()),
+    ])
+  );
+
+  return Yup.object().shape(lazyObjectShape).noUnknown();
+});
+
+/**
+ * User validation schema.
+ *
+ * @see User
+ */
+export const UserSchema: Yup.ObjectSchema<User> = Yup.object()
+  .shape<User>({
+    partyName: Yup.string(),
+    pictureUrl: Yup.string(),
+    anonMode: Yup.boolean(),
+    mirrorVideo: Yup.boolean(),
+    status: Yup.string(),
+    data: UserExperienceDataSchema,
+    myPersonalizedSchedule: MyPersonalizedScheduleSchema,
+    enteredVenueIds: Yup.array().of(Yup.string().required()),
+
+    // Legacy?
+    room: Yup.string(),
+    video: VideoStateSchema.notRequired(),
+    kidsMode: Yup.boolean(),
+  })
+  .noUnknown()
+  .required();
+
+// export const userWithLocationToUser = (
+//   user: WithId<UserWithLocation>
+// ): WithId<User> =>
+//   withId(UserSchema.validateSync(user, { stripUnknown: true }), user.id);
+
+// @debt Not sure if the validations are too 'heavyweight' for this, but object destructuring seemed to work
+//  here, whereas the validations seemed to hang my browser tab. There might also be something wrong with the
+//  validation rules leading to infinite recursion or similar?
+export const userWithLocationToUser = (
```

Comment:
Is this the correct file for this helper function? Also it seems to destructure the user and return `userWithoutLocation` but the name is `userWithLocationToUser`

Reply:
I put it here in a similar way to how we have the tiny `withId` helper alongside the `WithId` type.

My thinking was that having it 'close' to the types made most sense (and I think originally it was also using the validations from this file, so felt even more like it made sense)

I can see arguments for why it should be in somewhere like `utils/` or similar as well.. i'm kind of split 50/50 on it tbh.

---

> Also it seems to destructure the user and return `userWithoutLocation` but the name is `userWithLocationToUser`

The naming is correct, if not a little lacking in verbosity. It takes a `userWithLocation` and converts it `To` a `User`; thus `userWithLocation` `To` `User`.

This PR refactors the base `User` type to remove the location details from it, thus only the `UserWithLocation` type has them.

Reply:
@denisdimitrov IMO no changes are required here, so marking this as resolved.